_posts/2017-12-27-cheap-docker-swarm-cluster-aws-terraform.md

```diff
@@ -0,0 +1,98 @@
+
+When self-hosting a cluster, it’s not just about the cost of hardware, but also about the electricity used, which as we will find out can be significant.
+
+You could host your cluster using any of major cloud players out there, such as [Microsoft Azure](https://azure.microsoft.com/), [Google Cloud](https://cloud.google.com/compute/), or [IBM BlueMix](https://www.ibm.com/cloud/).  Each has different advantages, with comparable pricing.   But for this article I’m just going to focus on [Amazon Web Services](https://aws.amazon.com/).  Leave a comment below if you'd like us to compare cloud providers in a future article.
+
+## Self Hosting with Commodity Hardware
+
+So let’s start with our baseline.  What would it cost to get up and running with a 3 node cluster using generally available?  Just using parts from NewEgg, you can build a cheap commodity system for about $500.  Circa 2017, this would get you 4GB of RAM, and a 2.9GHz Dual Core Intel processor.  Most of that cost is overhead, such as power supply, case, motherboard, storage space, networking and cooling.  To make a good comparison with the cloud, let’s amortize the cost of the hardware over a 2 year period, which gives us an amortized rate of $20.83/month.
+
+But we also need to consider the cost of electricity.  Let’s assume a $0.10/kWh rate, which is quite inexpensive, but also common across large parts of North America.  Assuming a 300W constant power draw, we are looking at $21.60/month, just for electricity.  Our cost of electricity is about the same amount as our amortized hardware costs!  Not really what I had in mind for cheap, but this is just our baseline.
+
+Ignoring a large number of other considerable factors, this gives us an estimated monthly cost of operation of $42.43.  This is the amortized cost of the hardware, plus the cost of electricity.
+
+There are a number of alternatives we could explore on the self hosted side that could reduce our costs.  For example, we could explore the use of low power ARM processors, which are both inexpensive as well as efficient.  Depending on the intended purpose, this may be a valid option.  But for the purpose of comparison, we will stick with an Intel processor.
+
+To summarize:
+
+| Description | Cost |
+| ----------- | ---- |
+| Monthly Amortized Hardware Costs | $20.83 |
+| Monthly Electricity Usage        | $21.60 |
+| Total                            | $42.43 |
+
+## Cloud Hosting
+
+With [AWS](https://aws.amazon.com/), there are a large number of server types to choose from.  A comparable EC2 type to our self-hosted solution would be the t2.medium instance.  [Here’s a great site that compares the various options and prices.](https://www.ec2instances.info/)  If we use on-demand pricing, we will get higher availability, but we are paying full price.  We have the option to go with spot pricing and we are going to sacrifice availability for some savings.  Here’s the comparison for the t2.medium, with current 2017 pricing.
```

Comment:
Might be nice to also have a reference to spot pricing and what it means in terms of actual runtime performance. (And why are we not considering SNS/SQS/Lambda? ... is that even an option? Is it in scope? It may not be.)